### DISCUSSION TITLE
Enable Link-Time Optimization (LTO)

### DISCUSSION THREAD

**Discussion by zamazan4ik** [obi1kenobi/cargo-semver-checks#928] [Ideas]
Hi!

I noticed that in the `Cargo.toml` file Link-Time Optimization (LTO) for the project is not enabled. I suggest switching it on since it will reduce the binary size (always a good thing to have) and will likely improve the application's performance a bit (in this case, it's not critical at all but anyway).

I suggest enabling LTO only for the Release builds so as not to sacrifice the developers' experience while working on the project since LTO consumes an additional amount of time to finish the compilation routine. If you think that a regular Release build should not be affected by such a change as well, then I suggest adding an additional `dist` or `release-lto` profile where additionally to regular `release` optimizations LTO also will be added. Such a change simplifies life for maintainers and others interested in the project persons who want to build the most performant version of the application. Using ThinLTO also should help. If we enable it on the Cargo profile level, users, who install the application with `cargo install`, will get the LTO-optimized version "automatically".

Thank you.

**Comment by obi1kenobi** [comment]
Neat idea! Would you be open to creating a PR with a `release-lto` profile and making our binary building step (post-publish) use that new profile? I'd be happy to merge it!

**Reply by zamazan4ik** [reply]
I would be happy to do it for the project! However, I [have](https://github.com/zamazan4ik/awesome-pgo/blob/main/todo.md) a pretty big TODO list for my PGO efforts so... I don't want to promise anything falsely but I will try my best to make LTO enabled for the project :)

**Reply by obi1kenobi** [reply]
No worries! I've personally seen what PGO can do — I had 20% gains on one project earlier in my career — so I'd love to have both PGO and LTO eventually. It'd be nice to benchmark the difference so we know what we're getting, but we have an easy benchmark already and I'm not that worried about (moderate) increases to our precompiled binary's compilation time. If it takes us an extra minute to build the binary but our downstream users get 20% more perf, that's easily worth it.

So take your time, and we'd be happy to have your contributions anytime!
